### ISSUE TITLE
Latest: BDH-CQ: In-Context Learning with Recurrent Latent Reasoning

### ISSUE THREAD

**Issue by jjakimoto** [jjakimoto/research-issues#1174]
**Topic:** Quantitative trading; machine learning for quant & trading; agentic coding techniques; LLMs & multimodal LLMs; and autonomous research (autoresearch / AI-scientist) systems — methods, papers, tools  
**Date:** 2026-08-29

**Source:** huggingface-daily-papers · 2026-08-10
**Link:** https://arxiv.org/abs/2608.09888

A 150M-parameter closed model from Pathway hits 29.5% pass@2 on the *public* ARC-AGI-1 eval split for a self-computed $0.00070/task, claiming a new cost-efficiency SOTA without any test-time training or per-task optimization — but the headline is a cost-axis result, not an accuracy one, and the score and cost are both self-measured on a public split against leaderboard-reported baselines. Currently the top-trending paper on HF daily papers (761 upvotes).

## Executive summary

BDH-CQ combines in-context learning (demonstrations are written sequentially into a recurrent memory state) with recurrent latent reasoning (R iterations over a continuous workspace, decoded only at the end — no chain-of-thought tokens). A 150M-parameter configuration solves 118/400 tasks (29.5% pass@2) on the 400-task public ARC-AGI-1 evaluation split in ~0.85 H200 GPU-seconds per task, which the authors price at $0.00070/task and plot as breaking the ARC Prize cost-accuracy Pareto frontier. Unlike HRM/TRM, no parameters are updated at inference and no evaluation-task demonstration pairs or task identifiers appear in training. The bulk of the paper is a behavioral dissection on ConceptARC and post-freeze generated tasks that maps sharp, structured failure modes — but the model, its dimensions, update rules and training recipe are all proprietary, and the accuracy number is well below the leaderboard leaders it is priced against.

### The claim: a cost-axis SOTA, not an accuracy SOTA

The paper's headline is explicitly about the cost-accuracy plane, not absolute ARC performance. A 150M-parameter system reaches 29.5% pass@2 on the public ARC-AGI-1 evaluation split at a computed $0.00070 per task, a point the authors say no plotted leaderboard system matches at equal-or-lower reported cost. The comparison anchor is GPT-5.6 Luna (Low), which is *more* accurate (34.2%) but costs $0.040 per ARC Prize's July 2026 figures.

- 29.5% pass@2 = 118/400 tasks on the public ARC-AGI-1 evaluation split (not semi-private, not the ARC Prize verified set).
- Cost is computed as 0.85 H200 GPU-seconds x $3 per H200-hour = $0.00070/task — measured hardware time on their own infrastructure, not an API price.
- The paper is explicit that other plotted systems 'use the costs reported by the leaderboard, which may represent hardware estimates or API prices'.
- Claimed ~57x cheaper than GPT-5.6 Luna (Low); ~11x cheaper after OpenAI's 80% API price cut of July 30, 2026 that ARC Prize had not yet reflected.
- Reference points for the transductive competition: ARC Prize reports $1.48/task for HRM and $1.76/task for TRM.

### Architecture: recurrent memory + latent iteration, described only at interface level

BDH-CQ builds on the authors' earlier Dragon Hatchling (BDH) architecture — high-dimensional positive activations, low-rank communication, a recurrent associative state, ReLU-low-rank layers plus linear attention. Demonstrations are ingested sequentially into a recurrent state S_t with theta fixed, avoiding a growing KV cache. After the K demonstrations and the query are ingested, the query is encoded into a latent workspace H_0, iterated R times, and decoded once. Intermediate states are never decoded into language.

- Two distinct states with distinct roles: S_t accumulates in-context evidence; H_r carries the current query's computation.
- The recurrent-memory update is positioned as analogous to attention / fast-weight memory, with linear attention as the special case.
- Critically for reproducibility: the paper gives only interface equations. Dimensions, exact update rules, implementation details and the internal training recipe are all withheld as proprietary; no weights are released.
- The only public artifact is the ARC task generator repo (pathwaycom/arc-task-gen), not the model.

### Training setup: what was and was not trained on

A single 150M-parameter model is trained on curated ARC-style data: privately curated examples plus the ARC-AGI-1 training set, RE-ARC, ConceptARC, ARC-Heavy and ARC-GEN100K, with additional augmentations. The training objective is to produce exact target grids after preceding examples have been absorbed into recurrent context.

- Explicit claim, and the main methodological differentiator vs HRM/TRM: 'Neither task identifiers nor evaluation-task demonstration pairs participate in training, and no parameters are updated at inference time.' So: no test-time training, no per-puzzle optimization, no learned per-task identity embeddings.
- SKEPTIC'S FLAG: ConceptARC is in the training mixture (Section 4.2) and is simultaneously the primary behavioral-analysis benchmark in Section 6 (59.38% task pass@2). The authors concede their opaque-ID replication 'does not make ConceptARC a fresh benchmark, rule out exposure through training or checkpoint selection'.
- The ARC-AGI-1 *training* split is in the mixture (standard practice), while the scored 400-task *evaluation* split is held out.
- Reasoning effort is varied during training so that LOW/MEDIUM/HIGH latent-iteration budgets can be selected at inference.

### Evaluation and the cost-accuracy Pareto

Scoring follows the leaderboard's two-attempt pass@2 convention on the 400-task public evaluation split, with pass@1 and test-pair accuracy reported alongside. A black-box audit reproduced the 29.5% figure without access to weights — but it was performed by co-authors of this same paper, so it is a protocol replication rather than third-party evaluation.

- Task-level: pass@1 97/400 (24.25%), pass@2 118/400 (29.50%), Wilson 95% [25.24, 34.15] — the interval overlaps several nearby systems.
- Test-pair level (419 pairs): pass@1 25.78%, pass@2 31.03%.
- pass@2 is not always two independent attempts: in the opaque ConceptARC replication all 75 single-candidate records were already correct at rank one.
- Runs are deterministic: repeated identical requests were byte-identical across all 419 test inputs at both effort tiers.
- COST INCONSISTENCY: Section 6.6 prices the same STANDARD setting (also scoring 118/400) at $0.00265246 per task, ~3.8x the headline $0.00070 — the paper never reconciles the two figures.
- Table 5's 'cost reduction' for MEDIUM/LOW effort is only 11%/22%, which also does not line up with Section 6.6's claim that MIN effort costs one third of standard.

### Behavioral analysis: structured, sharp failure boundaries

The most substantive contribution is the diagnostic work. Post-freeze deterministic generators vary one complexity factor at a time, and ConceptARC provides a concept-organized profile. The picture is not a smooth difficulty scale but hard cliffs tied to demonstration coverage, output construction, and representation.

- Contextual binding is strong: a fresh per-task color permutation is applied correctly on all 96 held-out outputs at rank one, up to eight simultaneous bindings.
- Boundary propagation (48/48, distances 2-8) and motif copying (48/48, 1-4 sites) extrapolate without reaching a ceiling.
- Ordering collapses with length: near-saturated through five objects, 29/36 at six, 8/24 at seven, 1/24 at eight; at length eight only 3/24 outputs even have the correct dimensions.
- Adding ONE demonstration at the test complexity fixes depth-five nesting completely (19/24 -> 24/24 pass@2) and rescues ordering from 0/24 to 13/24 — so much of the 'cliff' is failure to extrapolate beyond demonstrated coverage.
- Composition is representation-dependent: rotation+relocation 72/72, reflection+relocation only 47/72, color swap 26/72 alone and 0/72 composed.
- Consistency gap: 52/160 ConceptARC tasks get one or two of three test pairs right — the inferred rule is not applied consistently.
- Appendix ladders: conditional rule selection drops 100% -> 56.7%; an undemonstrated parameter value drops 30% -> 0/120; panel union 2 -> 3 panels drops 65% -> 2.5%; support-chain depth 1 -> 8 drops 80% -> 27.5%.

### Significance and how far the claim actually reaches

If the numbers hold, the interesting result is not 29.5% — it is that a 150M-parameter model with no per-task optimization lands in the same accuracy neighborhood as systems costing three to four orders of magnitude more per task, which leaves headroom to scale. The forward-looking claims (1T-scale sharding, Transformer-like pretraining scaling from 1B to 600B parameters) are asserted from unreported 'early experiments' and carry no data in this paper. ARC-AGI-2 is named as future work only.

- The real contribution to a practitioner: evidence that in-context learning and latent (non-verbalized) reasoning can be combined in one recurrent fabric, with no inference-time gradient steps.
- The diagnostic methodology (post-freeze deterministic generators, opaque-ID replication, hand-written vs generated controls) is unusually careful and self-critical for a leaderboard-adjacent paper.
- But the model is closed, the score is on a public split, and the cost is self-computed — none of the three legs of the headline is externally auditable today.

**Table 1 — Headline results: public ARC-AGI-1 and ConceptARC**

| Set / condition | Unit | N | pass@1 | pass@2 [95% Wilson] |
| --- | --- | --- | --- | --- |
| ARC-AGI-1 public | tasks | 400 | 97 (24.25%) | 118 (29.50%) [25.24, 34.15] |
| ARC-AGI-1 public | test pairs | 419 | 108 (25.78%) | 130 (31.03%) |
| ConceptARC, semantic IDs | tasks | 160 | 73 (45.63%) | 95 (59.38%) [51.63, 66.68] |
| ConceptARC, opaque IDs | tasks | 160 | 72 (45.00%) | 96 (60.00%) [52.26, 67.27] |
| ConceptARC, opaque IDs | test pairs | 480 | 334 (69.58%) | 374 (77.92%) |

**Cost-accuracy comparison points cited in the paper (baselines are ARC Prize leaderboard-reported costs)**

| System | ARC-AGI-1 score | Cost per task | Basis / note |
| --- | --- | --- | --- |
| BDH-CQ (150M, standard effort) | 29.5% pass@2 (public 400-task split) | $0.00070 | Computed: 0.85 H200 GPU-s x $3/H200-hour, self-measured |
| BDH-CQ (same, per Section 6.6) | 118/400 pass@2 | $0.00265246 | Paper's own alternative figure for STANDARD effort — unreconciled with $0.00070 |
| BDH-CQ (MIN effort, Section 6.6) | 111/400 pass@2 (-1.75 pp) | $0.00088399 | McNemar p=0.167 vs standard — difference statistically unresolved |
| GPT-5.6 Luna (Low) | 34.2% | $0.040 | ARC Prize leaderboard, July 2026; ~57x more expensive (~11x after the July 30 2026 price cut) |
| HRM | not restated here | $1.48 | ARC Prize reported; transductive, requires per-task optimization |
| TRM | not restated here | $1.76 | ARC Prize reported; transductive, requires per-task optimization |

**Table 5 — pass@2 vs reasoning (latent thinking) effort**

| Effort | Pass@2 | Cost reduction |
| --- | --- | --- |
| HIGH | 29.5% | 0% |
| MEDIUM | 27% | 11% |
| LOW | 21% | 22% |

**Table 3 — Demonstration coverage rescues the extrapolation cliffs (24 byte-identical held-out inputs)**

| Family | Context | Output p@1 | Output p@2 |
| --- | --- | --- | --- |
| Ordering (length 8) | Short | 0/24 | 0/24 |
| Ordering (length 8) | Supported | 12/24 | 13/24 |
| Nesting (depth 5) | Short | 15/24 | 19/24 |
| Nesting (depth 5) | Supported | 16/24 | 24/24 |

## Key findings

- A 150M-parameter model reaches 29.5% pass@2 (118/400) on the PUBLIC ARC-AGI-1 evaluation split at a self-computed $0.00070/task — presented as a new cost-efficiency SOTA, not a new accuracy SOTA.
- No test-time training and no task-specific optimization: task identifiers and evaluation-task demonstration pairs are excluded from training, and no parameters are updated at inference — the explicit contrast with HRM/TRM, whose ARC pipelines are transductive and cost $1.48-$1.76 per task.
- Reasoning effort scales accuracy monotonically over three points (LOW 21%, MEDIUM 27%, HIGH 29.5%), but the reported cost savings (22%/11%/0%) are small and inconsistent with Section 6.6's claim that MIN effort costs one third of standard.
- Failure boundaries are sharp and coverage-driven: ordering collapses to 1/24 at length eight and nesting to 29/36 at depth five, but adding ONE demonstration at the test complexity restores nesting to 24/24 and ordering to 13/24 from 0/24.
- Composition is representation-dependent, not uniform: rotation composes with relocation 72/72 while reflection reaches only 47/72, and color swap is never composed successfully (0/72). Contextual binding, by contrast, is a genuine strength: 96/96 held-out color-permutation outputs correct at rank one.
- Rule consistency is weak: 52/160 ConceptARC tasks have one or two of three test pairs correct, so the model often produces correct outputs without applying a single consistent transformation.

## Notable quotes

> Rather than compressing the demonstrations into a single task vector, BDH-CQ processes their elements sequentially.
> At $3 per H200-hour this gives a computed cost of $0.00070 per task, less than one-tenth of a cent.
> Dimensions, exact update rules, and implementation details remain proprietary.
> Every stratified task was authored by GPT-5.6 from a prompt naming the target mechanic, so each rate also reflects how that generator instantiates the requested operation.

## Limitations

- THE HEADLINE IS NARROWER THAN IT SOUNDS AND SELF-MEASURED: 'state of the art' refers strictly to benchmark COST efficiency; 29.5% is below the compared frontier model (34.2%), the split is the PUBLIC 400-task eval run by the authors themselves, and the $0.00070 is a computed hardware figure on their own infrastructure while baselines are leaderboard costs that 'may represent hardware estimates or API prices'. ARC-AGI-2 is not attempted.
- INTERNAL COST CONTRADICTION: Section 6.6 states the STANDARD setting costs $0.00265246 per task for the same 118/400 result — 3.8x the headline — and the paper never reconciles it; Table 5's 11%/22% reductions also contradict Section 6.6's 'MIN effort cost one third of standard'.
- NOT REPRODUCIBLE AND POSSIBLY CONTAMINATED: dimensions, update rules and the training recipe are proprietary with no weights released (only a task generator); ConceptARC sits in the training mixture and is then the main behavioral benchmark; and the 'independent' black-box audit was run by co-authors, not a third party.

## Open questions

- Does the 29.5% survive on the ARC-AGI-1 semi-private / ARC Prize verified evaluation, where public-split contamination and checkpoint selection cannot help?
- Which cost figure is correct — $0.00070 or the $0.00265246 the paper reports for the same standard-effort 118/400 run — and how would like-for-like accounting (serving overhead, batching, utilization) compare against API-priced baselines?
- Is the recurrent-memory + latent-workspace design actually responsible for the efficiency, or is it the ARC-specialized 150M model plus the heavily augmented ARC-style training mixture? There is no ablation isolating architecture from data recipe.

---
_Generated by Auto Research via a headless coding agent. Sources were retrieved with live web search — please verify before citing._

**Comment by x711criptic** [comment]
If this is about agent tooling — x711 has 26 pay-per-use tools (web search, prices, shared memory, tx simulation) with a free tier. No signup to start, MCP supported.

```bash
curl -X POST https://x711.io/api/refuel \
  -H "Content-Type: application/json" \
  -d '{"tool":"price_feed","query":"ETH"}'
```

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.

**Comment by jjakimoto** [comment]
Superseded by a fresh Auto Research run — closing to reset the slate.
